--- request/wms/__init__.py.orig	2025-07-27 14:40:42.806807675 +0000
+++ request/wms/__init__.py	2025-07-27 14:41:12.013018858 +0000
@@ -186,7 +186,7 @@
     #pylint: disable-msg=E1102
     xml_exception_handler = None
     prevent_image_exception = False
-    dimension_params = ['time', 'elevation']
+    dimension_params = ['time', 'elevation', 'filter']
     dimension_prefix = 'dim_'
 
     def __init__(self, param=None, url='', validate=False, non_strict=False, **kw):
